File: project_euler/cpp/problem_3.cpp
/*
 * Project Euler problem template
 */

#include <chrono>
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

unsigned long problem_3(unsigned long input) {
    std::vector<unsigned long> primes;
    primes.reserve(10);  // Assume small number of primes

    const auto limit = std::sqrt(input) + 1;
    for (auto i = 2ul; i < limit; ++i) {

        if (input % i == 0) {

            // Divide i by all smaller primes of input to test for primeyness
            if (std::none_of(
                    primes.begin(), primes.end(),
                    [=](unsigned long const &p) { return 0 == i % p; })) {
                primes.push_back(i);
            }

        }

    }

    // Largest prime we found will be the larget prime divisor
    return primes.back();
}

int main(int argc, char *argv[]) {
    using namespace std::chrono;

    auto t1 = steady_clock::now();
    auto ans = problem_3(600851475143);
    auto t2 = steady_clock::now();

    auto time_span = duration_cast<duration<double>>(t2 - t1);

    std::cout << "Problem 3 answer = " << ans << " in " << time_span.count()
              << " seconds.\n";
    return 0;
}

File: project_euler/haskell/problem_3.hs
-- Project Euler Problem 3 find the largest prime factor of a number

-- divisors calculates all the divisors of a number not equal to 1 or itself.
divisors ::  Integral a  => a -> [a]
divisors num = concat [[x, div num x] | x <- [2..(limit-1)], num `mod` x == 0]
    where limit = (floor.sqrt.fromIntegral) num

-- isPrime determines if the input has any divisors other than 1 or itself
isPrime num | length (divisors num) == 0 = True
            | length (divisors num) > 0 = False

problem_3 :: Integral a => a -> [a]
problem_3 num = filter isPrime $ divisors num

main :: IO()
main = do
    y <- return $ maximum (problem_3 600851475143)
    putStrLn("Answer to problem 3 = " ++ show(y))
